### ISSUE TITLE
.env variables available locally but not in production bundle built in CodeBuild

### ISSUE THREAD

**Issue by GJordan904** [vuejs/vue-cli#4775]
### Version
4.0.5

### Environment info
```
Environment Info: 
 
  System: 
    OS: Linux 4.14 Ubuntu 18.04.3 LTS (Bionic Beaver) 
    CPU: (2) x64 Intel(R) Xeon(R) Platinum 8124M CPU @ 3.00GHz 
  Binaries: 
    Node: 10.16.3 - /usr/local/bin/node 
    Yarn: 1.17.3 - /usr/bin/yarn 
    npm: 6.9.0 - /usr/local/bin/npm 
  Browsers: 
    Chrome: 77.0.3865.90 
    Firefox: 69.0.1 
  npmPackages: 
    @types/vue-select: 2.5.0 => 2.5.0  
    @vue/babel-helper-vue-jsx-merge-props:  1.0.0  
    @vue/babel-plugin-transform-vue-jsx:  1.0.0  
    @vue/babel-preset-app:  4.0.5  
    @vue/babel-preset-jsx:  1.1.1  
    @vue/babel-sugar-functional-vue:  1.0.0  
    @vue/babel-sugar-inject-h:  1.0.0  
    @vue/babel-sugar-v-model:  1.1.1  
    @vue/babel-sugar-v-on:  1.1.0  
    @vue/cli-overlay:  4.0.5  
    @vue/cli-plugin-babel: 4.0.5 => 4.0.5  
    @vue/cli-plugin-e2e-cypress: 4.0.5 => 4.0.5  
    @vue/cli-plugin-pwa: 4.0.5 => 4.0.5  
    @vue/cli-plugin-router:  4.0.5  
    @vue/cli-plugin-typescript: 4.0.5 => 4.0.5  
    @vue/cli-plugin-unit-mocha: 4.0.5 => 4.0.5  
    @vue/cli-plugin-vuex:  4.0.5  
    @vue/cli-service: 4.0.5 => 4.0.5  
    @vue/cli-shared-utils:  4.0.5  
    @vue/component-compiler-utils:  3.0.0  
    @vue/preload-webpack-plugin:  1.1.1  
    @vue/test-utils: 1.0.0-beta.29 => 1.0.0-beta.29  
    @vue/web-component-wrapper:  1.2.0  
    typescript: 3.6.3 => 3.6.3  
    vue: 2.6.10 => 2.6.10  
    vue-class-component: 7.1.0 => 7.1.0  
    vue-hot-reload-api:  2.3.4  
    vue-js-modal: 1.3.31 => 1.3.31  
    vue-loader:  15.7.1  
    vue-property-decorator: 8.2.2 => 8.2.2  
    vue-router: 3.1.3 => 3.1.3  
    vue-select: 3.2.0 => 3.2.0  
    vue-style-loader:  4.1.2  
    vue-template-compiler: 2.6.10 => 2.6.10  
    vue-template-es2015-compiler:  1.9.1  
    vuex: 3.1.1 => 3.1.1  
    vuex-class: 0.3.2 => 0.3.2  
  npmGlobalPackages: 
    @vue/cli: 4.0.5 
```

### Steps to reproduce
When running the app locally with `vue-cli-service serve` I have access to my env variables that are defined in the .env files at the root of the project.

When I build for production in AWS CodeBuild, however, variables that I add during build time are not being picked up and injected into the bundle. The build is completing successfully though.

I have tried to build with the following commands, 
    - `npx vue-cli-service build --modern --mode $VUE_BUILD_MODE`
    - `npx vue-cli-service build --mode $VUE_BUILD_MODE` 

My initial attempt was to simply export the variables before building with `export VUE_APP_MY_VAR=xxx` this is what worked when using my own DotEnv implementation. I also tried adding the variables to the `.env` file and the `.env.$VUE_BUILD_MODE` file and then lastly the `.env.$VUE_BUILD_MODE.local` file. Although I created and appended my variables to those files, they are never picked up during the build.

The build always succeeds but all the VUE_APP_xxx variables are not available within the bundle.

### What is expected?
.env variables should be injected in both development and production.

### What is actually happening?
.env variables are only being injected in development

**Comment by GJordan904** [comment]
I figure I'll mention, while on the topic, all these various .env files are both excessive and in direct contradiction to the documentation for dotenv. The docs specifically state, "Do NOT use multiple environment specific .env files" and "Do not commit .env files". 

In my opinion, when building for production you should just read the system vars and not even look for a .env file, or give an option to turn off parsing the file.

**Comment by LinusBorg** [comment]
I am in no way familiar with AWS Codebuild but:

1.  A quick google shows that there are specific ways documented about how to set custom env variables in codebuild, e.g. [via the aws console(Point 12.) ](https://docs.aws.amazon.com/codebuild/latest/userguide/change-project.html)
2. AWS says this about secrets in env variables:

> We strongly discourage the use of environment variables to store sensitive values, especially AWS access key IDs and secret access keys. Environment variables can be displayed in plain text using tools such as the CodeBuild console and the AWS CLI.
>
> We recommend you store sensitive values in the Amazon EC2 Systems Manager Parameter Store and then retrieve them from your build spec. To store sensitive values, see Systems Manager Parameter Store and Systems Manager Parameter Store Console Walkthrough in the Amazon EC2 Systems Manager User Guide. To retrieve them, see the parameter-store mapping in Build Spec Syntax. 

And about your worries concerning our use of .env files:

> The docs specifically state, "Do NOT use multiple environment specific .env files" and "Do not commit .env files".

While correct, this has to be read in context, in my opinion. `dotenv` was written with node apps, i.e. an express server, in mind. In these contexts, environment vairabls would almost exclusively be used for things like database users and passwords - really important, secret and highly sensitive information.

In the context of frontend build processes though, environment variables are ver often used for information that will end up in the final bundle anyway, which means: it will be on the client, it will be more or less a 100% public and not really a secret. Usually, that stuff are API URLs, pubic API keys and similar stuff. In that context, we can argue that it makes total sense to commit such information about various environments, especially things like staging, to git.

And we do recognize that there's still situations where you have real secrets, i.e. where you need a secret, private API key during your build process to get some data from an API or whatever. for these kinds of secrets we have `.env.local` files which are matched by `.gitignore` so they won't be commited to git. And the recommendation is to use env variables provided by the system / CI chain when building for productron and not an .env file.

**Comment by GJordan904** [comment]
Thank you for the detailed response. For the sake of brevity I did not go into full details of how my CI setup works but maybe I should have. 

I use the Secret Manager for these particular environment variables. Essentially, during build time Codebuild gets some API tokens passed to it from Secret Manager. It is prior to calling the build command when I create the .env file on the build server but for some reason they are not being picked up.

If I run `cat .env` right before calling build, i will see that my variables have been added to the file. Likewise, if i simply export the variables and call `printenv` before building, I see the variables are in the systems environment but still not getting picked up. 

When I get back to the office I will be troubleshooting some more but I'm at a bit of a loss considering the variables are clearly present prior to build. The same variables work locally to when placed in .env.local file.

**Comment by LinusBorg** [comment]
Can you verify that the env variables are accessible inside of the node process?

i.e. do this inside `vue.config.js`:

```javascript
console.log(`Testing env variable VUE_BUILD_MODE: ${process.env.VUE_BUILD_MODE}`)
module.exports = {
  /// whatever
}
```

this will log the value of that env variable (whihc I understand you have set through the codebuild config as you explained). 

* If that fails, then then node seems to have trouble accessing this env variable provided by Codebuild
* If this succeeds, we seem to rather have somthing going on in vue-cli

**Comment by GJordan904** [comment]
Closing for I am a doofus sometimes. 

After closer inspection, there were some slight, single character, inconsistencies in variable names between my `.env.local` and the variables being passed into CodeBuild. Had I just checked `process.env` I would have seen them but I was looking for specific variables on the environment :man_facepalming: 

Thanks for working with me. I'm really liking the cli, great job :+1:
